src/snowflake/snowpark/_internal/ast_utils.py

```diff
@@ -133,6 +145,56 @@ def build_const_from_python_val(obj: Any, ast: proto.Expr) -> None:
         raise NotImplementedError("not supported type: %s" % type(obj))
 
 
+def build_fn_apply(
+    ast: proto.Expr,
+    builtin_name: str,
+    *args: Tuple[Union[proto.Expr, Any]],
+    **kwargs: Dict[str, Union[proto.Expr, Any]],
+) -> None:
+    """
+    Creates AST encoding for ApplyExpr(BuiltinFn(<builtin_name>, List(<args...>), Map(<kwargs...>))) for builtin
+    functions.
+    Args:
+        ast: Expr node to fill
+        builtin_name: Name of the builtin function to call.
+        *args: Positional arguments to pass to function.
+        **kwargs: Keyword arguments to pass to function.
+
+    Returns:
+
+    """
+
+    expr = ast.apply_expr
+
+    fn = proto.BuiltinFn()
+    fn.name = builtin_name
+    set_src_position(fn.src)
+    expr.fn.builtin_fn.CopyFrom(fn)
+
+    for arg in args:
+        if isinstance(arg, proto.Expr):
+            expr.pos_args.append(arg)
+        elif hasattr(arg, "_ast"):
+            assert arg._ast, f"Column object {arg} has no _ast member set."
+            expr.pos_args.append(arg._ast)
+        else:
+            pos_arg = proto.Expr()
+            build_const_from_python_val(arg, pos_arg)
+            expr.pos_args.append(pos_arg)
+
+    for name, arg in kwargs.items():
+        kwarg = proto.Tuple_String_Expr()
+        kwarg._1 = name
+        if isinstance(arg, proto.Expr):
+            kwarg._2.CopyFrom(arg)
```

Comment:
I'm curious about this line vs line 176.

At line 176, the code does `expr.pos_args.append(arg)` (uses `arg` as is).

Here, the code does `kwarg._2.CopyFrom(arg)`.

Is there a semantic difference between using the object reference as-is and cloning the object in depth? Would `CopyFrom` work equally well on line 176? (`expr.pos_args.append(Expr().CopyFrom(arg))` or some equivalent). Would `kwarg._2 = arg` work equally well here?

Reply:
I think line 176 is preferred, however I couldn't get line 189 working with a similar different syntax. `kwarg._2 = arg` gives an error, that's why I used the `CopyFrom`

Reply:
Thank you for clarifying. I still have memories from working on systems that differentiated between using a subtree in multiple places, `ShallowClone` and `DeepClone` 😃